src/v/cluster/tests/controller_forced_reconfiguration_test.cc

```diff
@@ -0,0 +1,424 @@
+            model::controller_ntp);
+        ASSERT_TRUE(maybe_controller_leader.has_value())
+          << "failed to get controller leader";
+        cluster::topic_properties props{};
+        props.cleanup_policy_bitflags = model::cleanup_policy_bitflags::none;
+        props.compaction_strategy = std::nullopt;
+
+        create_topic(
+          model::topic_namespace_view{
+            model::kafka_namespace, topic_config.name},
+          topic_config.partition_count,
+          topic_config.replication_factor,
+          std::move(props))
+          .get();
+    }
+
+    // restore the cluster to its original node size
+    void restore_node_number() {
+        scope_logger sl{"restore original node count"};
+        uint16_t first_survivor = _cluster_size / 2 + 1;
+
+        // for each dead node, create a replacement
+        for (const auto dead_node_number : typed_range(0, first_survivor)) {
+            const auto new_node_number = dead_node_number + _cluster_size;
+            create_node_application(
+              model::node_id{new_node_number}, model::node_id{first_survivor});
+        }
+
+        wait_for_all_members(small_wait).get();
+
+        // check that we have returned to _cluster_size from the perspective of
+        // every node
+        for (const auto node_number : typed_range(0, _cluster_size)) {
+            const model::node_id node_to_query{node_number + first_survivor};
+            // Wait for cluster to reach stable state
+            tests::cooperative_spin_wait_with_timeout(
+              medium_wait,
+              [this, node_to_query] {
+                  return get_local_cache(node_to_query).node_count()
+                         == _cluster_size;
+              })
+              .get();
+        }
+    }
+
+    // given a topic config, check that all partitions have leadership restored
+    // throws timeout exception otherwise
+    void wait_for_leader_restoration(topic_config topic_config) {
+        // step 1, wait for a controller leader
+        check_controller_leader();
+        auto [controller_rp_thread, controller_partition] = get_leader(
+          model::controller_ntp);
+
+        // step 2, force the leader to catch up
+        controller_rp_thread->app.controller->linearizable_barrier().get();
+
+        // step 3, find a leader for all partitions
+        for (const auto partition_number :
+             typed_range(0, topic_config.partition_count)) {
+            wait_for(large_wait, [this, &topic_config, partition_number] {
+                const auto ntp = model::ntp{
+                  model::kafka_namespace,
+                  topic_config.name,
+                  model::partition_id{partition_number}};
+                vlog(_logger.info, "polling for leadership on ntp: {}", ntp);
+                auto [rp_thread, partition] = this->get_leader(ntp);
+                return rp_thread != nullptr && partition != nullptr;
+            });
```

Comment:
wonder if this should also check all the the replicas are from the surviving et.